app/models/project_step.rb

```diff
@@ -10,299 +10,29 @@
+#  index_timeline_entries_on_agent_id                     (agent_id)
+#  index_timeline_entries_on_project_type_and_project_id  (project_type,project_id)
 #
 # Foreign Keys
 #
 #  fk_rails_a9dc5eceeb  (agent_id => people.id)
 #
 
 require 'chronic'
+class ProjectStep < TimelineEntry
+  class NoChildrenAllowedError < StandardError; end
 
-class ProjectStep < ActiveRecord::Base
-  include Translatable, OptionSettable
-
-  COLORS = {
-    on_time: "hsl(120, 73%, 57%)",
-    super_early: "hsl(120, 41%, 47%)",
-    barely_late: "hsl(56, 100%, 66%)",
-    super_late: "hsl(0, 74%, 54%)",
-  }
-  SUPER_EARLY_PERIOD = 7.0 # days
-  SUPER_LATE_PERIOD = 30.0 # days
-
-  default_scope { order('scheduled_date') }
-
-  belongs_to :project, polymorphic: true
-  belongs_to :agent, class_name: 'Person'
-
-  delegate :division, :division=, to: :project
-
-  has_many :project_logs, dependent: :destroy
-
-  attr_translatable :summary, :details
-
-  attr_option_settable :step_type
-
-  validates :project_id, presence: true
-  validate :unfinalize_allowed
-
-  before_save :handle_original_date_logic
-  before_save :handle_finalized_at
-
-  def division
-    project.try(:division)
-  end
-
-  def name
-    summary
-  end
-
-  # For use in views if record not yet saved.
-  def id_or_temp_id
-    @id_or_temp_id ||= id || "tempid#{rand(1000000)}"
-  end
-
-  def logs_count
-    project_logs.count
-  end
-
-  def set_completed!(date)
-    update_attribute(:completed_date, date)
-  end
-
-  def completed?
-    completed_date.present?
-  end
-
-  def completed_or_not
-    completed? ? 'completed' : 'incomplete'
-  end
-
-  def milestone?
-    self.step_type_value == "milestone" ? true : false
-  end
-
-  def last_log_status
-    project_logs.order(:date).last.try(:progress)
-  end
-
-  def admin_date_status
-    days = days_late
-    if days
-      if days <= 0
-        I18n.t('project_step.status.on_time')
-      else
-        I18n.t('project_step.status.days_late', days: days)
-      end
-    else
-      I18n.t(:none)
-    end
-  end
-
-  def admin_status
-    last_log_status || admin_date_status
-  end
-
-  def status
-    if completed?
-      I18n.t :log_completed
-    else
-      last_log_status
-    end
-  end
-
-  def display_date
-    I18n.l (self.completed_date || self.scheduled_date), format: :long
-  end
-
-  def date_changed?
-    original_date.present?
-  end
-
-  private
-
-  def handle_original_date_logic
-    # Note, "is_finalized" means a step is no longer a draft, and future changes should remember
-    # the original scheduled date.
-    if scheduled_date_changed? && is_finalized?
-      if original_date.blank?
-        self.original_date = scheduled_date_was
-      end
-      self.date_change_count = self.date_change_count.to_i.succ
-    end
-  end
-
-  def handle_finalized_at
-    if is_finalized && !finalized_at
-      self.finalized_at = Time.now
-    elsif !is_finalized && finalized_at
-      self.finalized_at = nil
-    end
-  end
-
-  public
```

Comment:
Please refactor so all the private methods are on bottom. Shouldn't be cycling back and forth like this.